File: models/World.ts
import VehicleController from '../controllers/VehicleController';
import Bullet from './Bullet';
import Vehicle from './Vehicle';
import Action from './Action';
import Heart from './Heart';
import { checkBoundariesOvelap } from '../helpers/math';
import { Position } from '..';

class World {

  private vehicles: WorldVehicles = {};
  private bullets: Array<Bullet> = [];
  private hearts: Array<Heart> = [];
  private vehicleControllers: Array<VehicleController> = [];
  private vehicleActionsOrder: Array<string> = [];
  private updateInterval: number|null = null;
  private executingVehicleActions: boolean = false;
  private updatingObjects: boolean = false;
  private width = 500;
  private height = 500;

  static ACTIONS = {
    ADD_BULLET: 'ADD_BULLET',
  };

  constructor (width: number = 500, height: number = 500) {
    this.width = width;
    this.height = height;
  };
  
  startUpdates() {
    this.updateInterval = setInterval(() => this.updateObjects(), 10);
  }

  addHeart(heart: Heart): void {
    this.hearts.push(heart)
  }

  addHeartOnRandomPosition(): void {
    this.addHeart(new Heart(
      (Math.random() * (this.width - 20)) + 10,
      (Math.random() * (this.height - 20)) + 10
    ))
  }

  addVehicleController (controller: VehicleController): void {
    controller.addListener((action: Action) => {
      this.getVehicle(controller.getVehicleId()).addAction(action);
      this.vehicleActionsOrder.push(controller.getVehicleId());
    });
    this.vehicleControllers.push(controller);
  }

  executeAction(action: Action, vehicle_id: string) {
    switch (action.getType()) {
      case World.ACTIONS.ADD_BULLET:
        this.bullets.push(new Bullet(
          action.getParam('x', 10),
          action.getParam('y', 10),
          action.getParam('angle', 0),
          vehicle_id,
          action.getParam('speed', 50),
        ))
        break;
      default:
        console.error("World action type not found", action.getType());
    }
  }

  executeVehicleActions () {
    if (this.executingVehicleActions) return;

    let currentActions = [ ...this.vehicleActionsOrder ];
    this.vehicleActionsOrder = [];

    this.executingVehicleActions = true;

    while (currentActions.length > 0) {
      const vehicle_id = currentActions.shift();
      if (!vehicle_id) throw new Error('Empty vehicle id!');

      const vehicle = this.getVehicle(vehicle_id);

      const result_action = vehicle.executeAction();

      if (result_action) {
        this.executeAction(result_action, vehicle_id);
      }
    }

    this.executingVehicleActions = false;
  }

  updateObjects () {
    this.executeLoops();

    if (!this.executingVehicleActions && this.vehicleActionsOrder.length > 0) {
      this.executeVehicleActions();
    }

    if (this.updatingObjects) return;

    this.updatingObjects = true;

    Object.values(this.vehicles).forEach( vehicle => {
      vehicle.update();
    });

    this.bullets.forEach(bullet => {
      bullet.update();
    });

    this.checkCollisions();

    this.updatingObjects = false;
  }

  checkCollisions() {
    Object.values(this.vehicles).forEach((vehicle) => {
      const vehicleBoundaries = vehicle.getBoundaries();
      this.bullets.forEach((bullet, k) => {
        const bulletBoundaries = bullet.getBoundaries();
        if (checkBoundariesOvelap(vehicleBoundaries, bulletBoundaries)) {
          vehicle.applyDamage(bullet.getDamageToApply(vehicle))
          this.bullets.splice(k, 1);
        }
      });
      this.hearts.forEach((heart, k) => {
        const heartBoundaries = heart.getBoundaries();
        if (checkBoundariesOvelap(vehicleBoundaries, heartBoundaries)) {
          vehicle.removeDamage(heart.getDamageToRemove(vehicle))
          this.hearts.splice(k, 1);
        }
      });
      if (vehicle.getLife() <= 0) {
        // TODO: Remove vehicleController
        delete this.vehicles[vehicle.getVehicleId()]
      }
    });

    this.bullets.forEach((bullet, k) => {
      const bulletPosition = bullet.getPosition();
      if (
        (bulletPosition.x < 0 - Bullet.RADIUS) ||
        (bulletPosition.y < 0 - Bullet.RADIUS) ||
        (bulletPosition.x > this.width + Bullet.RADIUS) ||
        (bulletPosition.y > this.height + Bullet.RADIUS)
      ) {
        return this.bullets.splice(k, 1);
      }

      const bulletBoundaries = bullet.getBoundaries();
      if (checkBoundariesOvelap(this.getBoundaries(), bulletBoundaries)) {
        return this.bullets.splice(k, 1);
      }
    });
  }

  executeLoops() {
    this.vehicleControllers.forEach(controller => {
      controller.loop();
    });
  }

  addVehicle (vehicle: Vehicle): string {
    let randomId = Math.random().toString(25);
    while (this.vehicles[randomId]) {
      randomId = Math.random().toString(25);
    }
    this.vehicles[randomId] = vehicle;
    vehicle.setVehicleId(randomId)
    return randomId;
  }

  getVehicle (id: string): Vehicle {
    return this.vehicles[id];
  }

  getVehicles (): Array<Vehicle> {
    return Object.values(this.vehicles);
  }

  addBullet (bullet: Bullet): void {
    this.bullets.push(bullet);
  }

  getBullets (): Array<Bullet> {
    return this.bullets;
  }

  getHearts (): Array<Heart> {
    return this.hearts;
  }

  getWidth (): number {
    return this.width;
  }

  getHeight (): number {
    return this.height;
  }

  getBoundaries (): Array<Position> {
    return [
      { x: 0, y: 0 },
      { x: this.width, y: 0 },
      { x: this.width, y: this.height },
      { x: 0, y: this.height },
    ];
  }
}

interface WorldVehicles {
  [key: string]: Vehicle
}

export default World

File: rust/models/World.rs
use crate::controllers::VehicleController;
use crate::models::{Bullet, Vehicle, Action, Heart};
use crate::helpers::math::check_boundaries_overlap;
use crate::interfaces::Position;
use std::collections::HashMap;

pub struct World {
    fps_counter: FPSCounter,
    vehicles: WorldVehicles,
    bullets: Vec<Bullet>,
    hearts: Vec<Heart>,
    vehicle_controllers: Vec<VehicleController>,
    vehicle_actions_order: Vec<String>,
    update_interval: Option<std::time::Duration>,
    executing_vehicle_actions: bool,
    updating_objects: bool,
    width: f64,
    height: f64,
}

impl World {
    pub const ACTIONS: [&'static str; 1] = ["ADD_BULLET"];

    pub fn new(width: f64, height: f64) -> Self {
        World {
            fps_counter: FPSCounter::new(),
            vehicles: HashMap::new(),
            bullets: Vec::new(),
            hearts: Vec::new(),
            vehicle_controllers: Vec::new(),
            vehicle_actions_order: Vec::new(),
            update_interval: None,
            executing_vehicle_actions: false,
            updating_objects: false,
            width,
            height,
        }
    }

    pub fn start_updates(&mut self) {
        self.stop_updates();
        self.update_interval = Some(std::time::Duration::from_millis(10)); // Placeholder for actual timer
    }

    pub fn stop_updates(&mut self) {
        self.update_interval = None;
    }

    pub fn add_heart(&mut self, heart: Heart) {
        self.hearts.push(heart);
    }

    pub fn add_heart_on_random_position(&mut self) {
        self.add_heart(Heart::new(
            rand::random::<f64>() * (self.width - 20.0) + 10.0,
            rand::random::<f64>() * (self.height - 20.0) + 10.0,
        ));
    }

    pub fn remove_heart(&mut self, index: usize) -> Heart {
        self.hearts.remove(index)
    }

    pub fn add_vehicle_controller(&mut self, controller: VehicleController) {
        controller.set_action_listener(|action| {
            if let Some(vehicle) = self.get_vehicle(controller.get_vehicle_id()) {
                vehicle.add_action(action);
                self.vehicle_actions_order.push(controller.get_vehicle_id());
            }
        });
        controller.set_world(self);
        self.vehicle_controllers.push(controller);
    }

    pub fn execute_action(&mut self, action: Action, vehicle_id: &str) {
        match action.get_type() {
            "ADD_BULLET" => {
                self.bullets.push(Bullet::new(
                    action.get_param("x", 10.0),
                    action.get_param("y", 10.0),
                    action.get_param("angle", 0.0),
                    vehicle_id.to_string(),
                    action.get_param("speed", 50.0),
                ));
            }
            _ => {
                eprintln!("World action type not found: {}", action.get_type());
            }
        }
    }

    pub fn execute_vehicle_actions(&mut self) {
        if self.executing_vehicle_actions {
            return;
        }
        let current_actions = self.vehicle_actions_order.clone();
        self.vehicle_actions_order.clear();
        self.executing_vehicle_actions = true;
        for vehicle_id in current_actions {
            if let Some(vehicle) = self.get_vehicle(&vehicle_id) {
                if let Some(result_action) = vehicle.execute_action() {
                    self.execute_action(result_action, &vehicle_id);
                }
            }
        }
        self.executing_vehicle_actions = false;
    }

    pub fn update_objects(&mut self) {
        self.execute_loops();
        if !self.executing_vehicle_actions && !self.vehicle_actions_order.is_empty() {
            self.execute_vehicle_actions();
        }
        if self.updating_objects {
            return;
        }
        self.updating_objects = true;
        for vehicle in self.vehicles.values_mut() {
            vehicle.update();
        }
        for bullet in self.bullets.iter_mut() {
            bullet.update();
        }
        self.check_collisions();
        self.fps_counter.add_tick_to_current_second();
        self.updating_objects = false;
    }

    pub fn get_fps_last_second(&self) -> f64 {
        self.fps_counter.get_last_second_counter()
    }

    pub fn check_collisions(&mut self) {
        for vehicle in self.vehicles.values_mut() {
            let vehicle_boundaries = vehicle.get_boundaries();
            for (k, bullet) in self.bullets.iter_mut().enumerate() {
                let bullet_boundaries = bullet.get_boundaries();
                if check_boundaries_overlap(&vehicle_boundaries, &bullet_boundaries) {
                    vehicle.apply_damage(bullet.get_damage_to_apply(vehicle));
                    self.remove_bullet(k);
                }
            }
            for (k, heart) in self.hearts.iter_mut().enumerate() {
                let heart_boundaries = heart.get_boundaries();
                if check_boundaries_overlap(&vehicle_boundaries, &heart_boundaries) {
                    vehicle.remove_damage(heart.get_damage_to_remove(vehicle));
                    self.remove_heart(k);
                }
            }
            if vehicle.get_life() <= 0 {
                self.remove_vehicle(vehicle.get_vehicle_id());
            }
        }
        for (k, bullet) in self.bullets.iter_mut().enumerate() {
            let bullet_position = bullet.get_position();
            if bullet_position.x < 0.0 - Bullet::RADIUS || bullet_position.y < 0.0 - Bullet::RADIUS || bullet_position.x > self.width + Bullet::RADIUS || bullet_position.y > self.height + Bullet::RADIUS {
                self.remove_bullet(k);
            }
            let bullet_boundaries = bullet.get_boundaries();
            if check_boundaries_overlap(&self.get_boundaries(), &bullet_boundaries) {
                self.remove_bullet(k);
            }
        }
    }

    pub fn execute_loops(&mut self) {
        for controller in &mut self.vehicle_controllers {
            if let Some(vehicle) = self.vehicles.get(controller.get_vehicle_id()) {
                controller.loop();
            }
        }
    }

    pub fn add_vehicle(&mut self, vehicle: Vehicle) -> String {
        let mut random_id = rand::random::<u64>().to_string();
        while self.vehicles.contains_key(&random_id) {
            random_id = rand::random::<u64>().to_string();
        }
        self.vehicles.insert(random_id.clone(), vehicle);
        vehicle.set_vehicle_id(random_id.clone());
        vehicle.set_max_boundaries(self.width, self.height);
        random_id
    }

    pub fn remove_vehicle(&mut self, vehicle_id: &str) -> Option<Vehicle> {
        self.vehicles.remove(vehicle_id)
    }

    pub fn get_vehicle(&self, id: &str) -> Option<&Vehicle> {
        self.vehicles.get(id)
    }

    pub fn get_vehicles(&self) -> Vec<&Vehicle> {
        self.vehicles.values().collect()
    }

    pub fn add_bullet(&mut self, bullet: Bullet) {
        self.bullets.push(bullet);
    }

    pub fn remove_bullet(&mut self, index: usize) -> Bullet {
        self.bullets.remove(index)
    }

    pub fn get_bullets(&self) -> Vec<&Bullet> {
        self.bullets.iter().collect()
    }

    pub fn get_hearts(&self) -> Vec<&Heart> {
        self.hearts.iter().collect()
    }

    pub fn get_width(&self) -> f64 {
        self.width
    }

    pub fn get_height(&self) -> f64 {
        self.height
    }

    pub fn get_boundaries(&self) -> Vec<Position> {
        vec![
            Position { x: 0.0, y: 0.0 },
            Position { x: self.width, y: 0.0 },
            Position { x: self.width, y: self.height },
            Position { x: 0.0, y: self.height },
        ]
    }
}

pub type WorldVehicles = HashMap<String, Vehicle>;
